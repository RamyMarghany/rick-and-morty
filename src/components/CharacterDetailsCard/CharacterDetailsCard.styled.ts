import styled from "styled-components";

export const StyledMainDetails = styled.div`
    display: flex;
    flex-direction: row;
    text-align: left;
    gap: ${(props) => props.theme.spacing.medium};
    box-shadow: 0px 1px 5px 1px ${(props) => props.theme.colors.primaryText};
    padding: ${(props) => props.theme.spacing.medium};
    margin: ${(props) => props.theme.spacing.default} auto;
    & img{
        border-radius: ${(props) => props.theme.spacing.small};
        max-width: 100%;
        height: auto;
    }
    & b{
        display:block;
        margin-bottom: 0.5rem;
    }

    @media (max-width: 768px) {
        flex-direction: column;
        text-align: center;
        
        & img{
            border-radius: 50%;
            padding: ${(props) => props.theme.spacing.medium};
        }
    }
`