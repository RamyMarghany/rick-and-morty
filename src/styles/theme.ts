export const LightTheme = {
    colors: {
        background: "#FAFAFA",
        primaryText: "#202020",
        secondaryText: "#616161",
        primary: "#FF5252",
        secondary: "#536DFE",
        error: "#D32F2F"
    },
    spacing: {
        maxWidth: "1200px",
        small: "5px",
        medium: "10px",
        default: "20px",
        large: "40px",
    },
    typography: {
        xxsmall: "10px",
        xsmall: "12px",
        small: "14px",
        default: "16px",
        large: "20px",
        xlarge: "28px",
        xxlarge: "34px",
    },
};
export const DarkTheme = {
    ...LightTheme,
    colors: {
        background: "#2B2B2B",
        primaryText: "#E8E6E3",
        secondaryText: "#B3B3B3",
        primary: "#FF4081",
        secondary: "#448AFF",
        error: "#F44336"
    },
};