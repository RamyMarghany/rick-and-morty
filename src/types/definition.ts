export type CharactersType = {
    id: string,
    name: string,
    image: string,
    species: string,
}

export type EpisodesType = {
    id: string,
    name: string,
    created: string,
}