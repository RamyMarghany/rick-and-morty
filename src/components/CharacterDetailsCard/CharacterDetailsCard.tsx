import { Image } from "components/Image/Image";
import { CharacterDetailsType } from "types/definition";
import { StyledMainDetails } from "./CharacterDetailsCard.styled";
export const CharacterDetailsCard: React.FC<{
  characterDetails: CharacterDetailsType;
}> = ({ characterDetails }) => {
  return (
    <StyledMainDetails>
      <Image
        src={characterDetails.image}
        alt={characterDetails.name}
        width={300}
        height={300}
      />
      <div>
        <h3>{characterDetails.name}</h3>
        <p>
          Species<b>{characterDetails.species}</b>
        </p>
        <p>
          Gender<b>{characterDetails.gender}</b>
        </p>
        <p>
          Status<b>{characterDetails.status}</b>
        </p>
        <p>
          Location<b>{characterDetails.location.name}</b>
        </p>
        <p>
          Dimension<b>{characterDetails.location.dimension}</b>
        </p>
      </div>
    </StyledMainDetails>
  );
};
