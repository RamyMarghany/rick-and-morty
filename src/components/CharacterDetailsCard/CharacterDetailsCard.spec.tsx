import { render, screen } from "@testing-library/react";
import { CharacterDetailsCard } from "./CharacterDetailsCard";
import { CharacterDetailsType } from "types/definition";
import { BrowserRouter as Router } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LightTheme } from "styles/theme";

describe("CharacterDetailsCard", () => {
  const mockDetails: CharacterDetailsType = {
    id: "1",
    name: "John Doe",
    image: "path/to/image",
    species: "Human",
    gender: "Male",
    status: "Alive",
    location: {
      name: "Earth",
      dimension: "Dimension C-137",
      type: "Planet",
    },
    episode: [],
  };

  it("renders without crashing", () => {
    render(
      <Router>
        <ThemeProvider theme={LightTheme}>
          <CharacterDetailsCard characterDetails={mockDetails} />
        </ThemeProvider>
      </Router>
    );
    expect(screen.getByText("John Doe")).toBeInTheDocument();
  });

  it("passes correct props to Image component", () => {
    render(
      <Router>
        <ThemeProvider theme={LightTheme}>
          <CharacterDetailsCard characterDetails={mockDetails} />
        </ThemeProvider>
      </Router>
    );
    const image = screen.getByRole("presentation");
    expect(image).toHaveAttribute("src", "path/to/image");
    expect(image).toHaveAttribute("alt", "John Doe");
  });

  it("displays the correct character details", () => {
    render(
      <Router>
        <ThemeProvider theme={LightTheme}>
          <CharacterDetailsCard characterDetails={mockDetails} />
        </ThemeProvider>
      </Router>
    );
    expect(screen.getByText("Human")).toBeInTheDocument();
    expect(screen.getByText("Male")).toBeInTheDocument();
    expect(screen.getByText("Alive")).toBeInTheDocument();
    expect(screen.getByText("Earth")).toBeInTheDocument();
    expect(screen.getByText("Dimension C-137")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Router>
        <ThemeProvider theme={LightTheme}>
          <CharacterDetailsCard characterDetails={mockDetails} />
        </ThemeProvider>
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
