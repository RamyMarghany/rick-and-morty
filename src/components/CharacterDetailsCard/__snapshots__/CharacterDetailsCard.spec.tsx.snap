// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetailsCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-beySPh hrySto"
  >
    <picture>
      <source
        srcset="
            /path/to/image.jpeg?w=100 100w, 
            /path/to/image.jpeg?w=200 200w, 
            /path/to/image.jpeg?w=400 400w, 
            /path/to/image.jpeg?w=800 800w"
        type="path/to/image.jpeg"
      />
      <img
        alt="John Doe"
        decoding="async"
        height="300"
        loading="lazy"
        role="presentation"
        sizes="(max-width:800px) 100vw, 50vw"
        src="path/to/image"
        width="300"
      />
    </picture>
    <div>
      <h3>
        John Doe
      </h3>
      <p>
        Species
        <b>
          Human
        </b>
      </p>
      <p>
        Gender
        <b>
          Male
        </b>
      </p>
      <p>
        Status
        <b>
          Alive
        </b>
      </p>
      <p>
        Location
        <b>
          Earth
        </b>
      </p>
      <p>
        Dimension
        <b>
          Dimension C-137
        </b>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
